tiledb/api/src/column.rs

```diff
@@ -0,0 +1,416 @@
+                sized data. Use one of the specialized \
+                to_type_vec() methods.",
+            ));
+        }
+
+        // Check that we have the right number of bytes for the given datatype.
+        if self.data().len() % datatype.size() as usize != 0 {
+            return Err(Error::from("Invalid datatype for column data."));
+        }
+
+        // Check that the number of elements satisfies the cell value size.
+        let elems = self.data().len() / datatype.size() as usize;
+        if elems % cell_val_size as usize != 0 {
+            return Err(Error::from(
+                "Invalid cell value size for column data.",
+            ));
+        }
+
+        // At this point we're ready to create our Vec to return. For now
+        // I'm going to cowboy it a bit and assume that alignment is
+        // correct.
+        let vals: &[T] = unsafe {
+            std::slice::from_raw_parts(
+                self.data().as_ptr() as *const T,
+                self.data().len() / std::mem::size_of::<T>(),
+            )
+        };
+
+        Ok(vals.to_vec())
+    }
+
+    pub fn to_string_vec(
+        &self,
+        datatype: Datatype,
+        cell_val_size: u32,
+    ) -> TileDBResult<Vec<String>> {
+        // Currently we're only supporting ASCII and UTF-8 strings.
+        if !matches!(datatype, Datatype::StringAscii | Datatype::StringUtf8) {
+            return Err(Error::from(format!(
+                "Invalid datatype '{}' for to_string_vec",
+                datatype.to_string().unwrap_or("<unknown type>".to_owned())
+            )));
+        }
+
+        // For var sized data, use the to_type_vec methods for conversion.
+        if cell_val_size != u32::MAX {
+            return Err(Error::from("Use to_vec<T> for fixed sized data."));
+        }
+
+        if self.offsets().is_empty() {
+            return Ok(Vec::new());
+        }
+
+        let mut ret: Vec<String> = Vec::new();
+
+        let data = self.data();
+        let offsets = self.offsets();
+        let num_offsets = offsets.len();
+        for (idx, offset) in self.offsets().iter().enumerate() {
+            let len = if idx == num_offsets - 1 {
+                self.data().len() as u64 - *offset
+            } else {
+                offsets[idx + 1] - offset
+            };
+
+            let val = String::from_utf8(
+                data[(*offset as usize)..((*offset + len) as usize)].to_vec(),
+            )
+            .map_err(|e| {
+                Error::from(format!(
+                    "Invalid UTF-8 in column data. \
+                    Use to_boxed_u8_vec instead: {}",
+                    e
+                ))
+            })?;
+
+            ret.push(val);
+        }
+
+        Ok(ret)
+    }
+
+    pub fn to_boxed_u8_vec(
+        &self,
+        datatype: Datatype,
+        cell_val_size: u32,
+    ) -> TileDBResult<Vec<Box<[u8]>>> {
+        // Currently we're only supporting ASCII and UTF-8 strings.
```

Comment:
It looks like UTF-16 has decent support. It looks like we'd have to write our own UTF-32/UCS-2/UCS-4 converters based on the standard library. And/or just add a dependency if we can find something specific enough to be worth the tradeoff there.

Reply:
Since this is the box o' bytes version though we don't really need to deal with the encoding, it's up to whoever the user is going to be.  The line below this is `matches!(datatype, Datatype::StringAscii | Datatype::StringUtf8)` and that can be more permissive to allow the other types.

Reply:
That bit is mostly just convenience coupled with the fact that I wasn't thinking too hard since the `to_string_vec` is Generic I was avoiding dealing with more than one `String::from_whatever` calls. Though I could easily add the StringUtf16 variant using conditional logic.

Although, I think your underlying point is more "We should let folks have access to the raw data" which is technically possible with the `to_boxed_u8_vec`, but I could see there being a reasonable request for having `to_u8_slices` to avoid copies for folks wanting to do other things efficiently.

Reply:
Yes, that's my underlying point - no need to restrict the raw access.  This function, and a `to_u8_slices` if we add one, should allow any of the `StringWhatever` types